Tkinter

eyezhick opened the pull request:
Added main.py (temporary naming scheme).
Using Tk interface, created the Tk GUI for the Featurizer class.

minor modifications to Featurizer.py:
* Whenever a string is printed, it instead stores the output in a string object.
* added dot_to_png function that converts a gv file into a png image file (render part of the graphviz package).

eyezhick committed:
Tkinter

eyezhick committed:
deleted unneed files

eyezhick committed:
added GUI functionality

eyezhick committed:
sample output images

A folder for saving output images.

eyezhick committed:
Update Featurizer.py

connormayer commented on code/Featurizer.py:
```diff
@@ -1,9 +1,13 @@
 import argparse
 
+## addition
+#import pydot
```
You can completely remove these bits you've commented out.

connormayer commented on code/Featurizer.py:
```diff
@@ -1,9 +1,13 @@
 
+## addition
+#import pydot
+from graphviz import render
```
Can you move this down so it's alphabetical with the other from ... import.. statements? So between lines 9 and 10 here.

connormayer commented on code/Featurizer.py:
```diff
@@ -149,6 +153,23 @@ def calculate_class_features(self):
+
+        ## method using the pydot
+        #(graph,) = pydot.graph_from_dot_file(filename)
+        #graph.write_png('test.png')
```
You can remove these commented out lines.

connormayer commented on code/main.py:
```diff
@@ -0,0 +1,281 @@
+import Featurizer
+from tkinter import *
```
You might not need this line given the imports below? It's generally bad practice to do `import *` because it can result in unexpected namespace clashes.

eyezhick replied:
Instead of using the wildcard, I specified the namespace with 
```python  
import tkinter as tk
```
I'm not totally sure why, but the other imports statement seems necessary.
```python
from tkinter import ttk
from tkinter import filedialog
from tkinter import messagebox
```

The python manual says that I need something like this
```python
from tkinter import *
from tkinter.ttk import *
```
to overide the basic tk widgets with ttk, and the same goes for filedialog and messagebox.
```python
from tkinter import * 
from tkinter import messagebox 
```

connormayer commented on code/main.py:
```diff
@@ -0,0 +1,281 @@
+        self.title("Featurizer")
+        self.minsize(window_width, window_height)
+        #self.wm_iconbitmap(())
+        #self.configure(background='#4D4D4D')
```
Do these need to be kept?

connormayer commented on code/main.py:
```diff
@@ -0,0 +1,281 @@
+        #################################################
+        ##       RADIO BUTTON - FEATURIZATION TYPE     ##
+        #################################################
+        MODES = [
```
Can you change the order of these so that `privative` and `complementary` are swapped? This lines it up with how it's presented in the paper.

connormayer commented on code/main.py:
```diff
@@ -0,0 +1,281 @@
+
+    def input_validation(self):
+        ''' Checks if input filename has been selected '''
+        if self.filename_input is "":
```
You should use either `if self.filename_input == ""` or, even better, `if not self.filename_input`. `is` checks if two objects are the same object, while `==` checks if two objects have the same values. It particularly doesn't make sense to use `is` here because one of the arguments is the string literal `""`.

connormayer commented on code/main.py:
```diff
@@ -0,0 +1,281 @@
+        #self.configure(background='#4D4D4D')
+
+        # center the window
+        positionRight = int((self.winfo_screenwidth() / 2) - (window_width / 2))
```
Can you switch all the variable names in this file (aside from class names) to use snake case? So here, `position_right`.

connormayer commented on code/main.py:
```diff
@@ -0,0 +1,281 @@
+        #################################################
+        ##          RADIO BUTTON - SELECT TYPE         ##
+        #################################################
+        self.v = IntVar(None, 2)
```
I think rather than having the two "Select type" buttons, it might make more sense to replace the "PNG" and "GV" buttons in the "Save a file" section with buttons that are "Save poset graph as PNG", "Save featurization graph as PNG", "Save poset graph as GV", and "Save featurization graph as GV". The reason for this is that selecting poset or featurization doesn't change the CSV that gets output, which might be a bit confusing.

connormayer commented on code/main.py:
```diff
@@ -0,0 +1,281 @@
+        button = ttk.Button(self.labelFrame5, text="PNG", command=self.fileDialogSave_PNG)
+        button.grid(column=1, row=0)
+
+    def fileDialogSave_PNG(self):
```
Both this method and the one that saves the GV files should append to the end of the filename indicating whether it's a featurzation or poset plot/GV file. So something like, e.g., `toy_system_privative_poset.png` or `toy_system_privative_featurization.png'.

connormayer commented on code/main.py:
```diff
@@ -0,0 +1,281 @@
+        ''' Checks if input filename has been selected '''
+        if self.filename_input is "":
+            messagebox.showinfo("Error", "Please select an input file")
+            return True
```
It seems a bit weird that this method returns `True` if the input is invalid...

connormayer commented on code/main.py:
```diff
@@ -0,0 +1,281 @@
+        elif self.v.get() == 2:
+            featurizer.graph_feats(filename_temp)
+
+        featurizer.dot_to_png(filename_temp, filename_output)
```
It's possible for this call to fail even if the user has the Python `graphviz` library installed if they don't have graphviz installed. This raises a `graphviz.backend.ExecutableNotFound` error. It would be good to catch this and print a message telling them they need to install graphviz, maybe with a link to the graphviz website?

connormayer commented on code/main.py:
```diff
@@ -0,0 +1,281 @@
+        text_print = verbose_text + featurizer.print_featurization_helper() + \
+                     featurizer.print_segment_features_helper()
+
+            # A Label widget to show in toplevel
```
Remove this extra indent

eyezhick committed:
1st adjustment

Revised main.py and Featurizer.py based on the comments left on previous pull request

eyezhick committed:
2nd adjustment

connormayer commented on code/.idea/.gitignore:
```diff
@@ -0,0 +1,3 @@
+# Default ignored files
```
Can you add the .idea folder to the top level .gitignore? I don't think this needs to be part of the repo, right?

connormayer commented on code/main.py:
```diff
@@ -0,0 +1,285 @@
+import tkinter as tk
+from tkinter import ttk
+from tkinter import filedialog
+from tkinter import messagebox
```
Is there a reason to not just use `tk.filedialog` and `tk.messagebox`?

eyezhick replied:
I tried that at first and it did not work. After doing some research, I found out that both filedialog and messagebox are not included in the list of ```__all__``` of tkinter and that separate imports are required for those modules. They say that that's simply how the package was designed. Looking at how other more advanced tkinter users have done it, separate imports like so seem to be the standard with tkinter (I think).

connormayer replied:
Ok, fair enough!

connormayer commented on code/main.py:
```diff
@@ -0,0 +1,285 @@
+        button = ttk.Button(self.labelFrame, text="Browse", command=self.fileDialog)
+        button.grid(column=1, row=0)
+
+    def fileDialog(self):
```
Can you switch this method to snake case too? `file_dialog`. Ditto with the other fileDialog methods below.

connormayer commented on code/main.py:
```diff
@@ -0,0 +1,285 @@
+
+        try:
+            featurizer.dot_to_png(filename_temp, filename_output)
+        except:
```
Can you switch this to catch the `graphviz.backend.ExecutableNotFound` error specifically? What you've got here is called 'Pokemon exception handling' (gotta catch 'em all). This is sometimes ok, but in this case it's dangerous because it will report a missing graphviz executable for EVERY exception, not just those that are triggered by the missing executable. If the issue was something like file permissions, this would be very misleading.

I'd also expand the first error message below a little bit to say something like "Missing graphviz executable" just to be a bit more concrete.

connormayer commented:
Can you remove the generated .png files from your commits?

eyezhick committed:
3rd Adjustments

Deleted png folder
Added .idea to .gitignore
Specified exception handling
Modified method names (snakecase
